aspnetcore/security/key-vault-configuration.md

```diff
@@ -0,0 +1,144 @@
+
+```csharp
+public class EnvironmentSecretManager : IKeyVaultSecretManager
+{
+    private readonly string _appNamePrefix;
+
+    public EnvironmentSecretManager(string appName)
+    {
+        _appNamePrefix = appName + "-";
+    }
+
+    public bool Load(SecretItem secret)
+    {
+        return secret.Identifier.Name.StartsWith(_appNamePrefix);
+    }
+
+    public string GetKey(SecretBundle secret)
+    {
+        return secret.SecretIdentifier.Name.Substring(_appNamePrefix.Length);
+    }
+}
+```
+
+```csharp
+builder.AddAzureKeyVault(
+    $"https://{config["Vault"]}.vault.azure.net/",
+    config["ClientId"],
+    config["ClientSecret"],
+    new EnvironmentSecretManager(env.ApplicationName));
+    
+Configuration = builder.Build();
+```
+
+The `Load()` method is called by an algorithm in the provider that iterates through the secrets to find the one that matches the application name as a prefix to the secret's name. When a match is found with `Load()`, the algorithm uses the `GetKey()` method to return the configuration name of the secret name. It simply strips off the application name prefix from the secret's name and returns the name for loading into the app's configuration name-value pairs.
+
+If you implemented this approach with the sample application, the key vault secrets would be loaded, the string secret for `KeyVaultConfigProviderSample-ConnectionString` would be matched, and the application name `KeyVaultConfigProviderSample` (with the dash) would be stripped off to load `ConnectionString` with its value into the app's configuration.
+
+You can also provide your own `KeyVaultClient` implementation to `AddAzureKeyVault()`. Supplying a custom client allows you to share a single instance of the client between the configuration provider and other parts of your application.
+
+## Controlling access to the ClientSecret
+To maintain the `ClientSecret` outside of your project tree, you can use the [Secret Manager tool](xref:security/app-secrets), whereby you can associate app secrets with a specific project and share them across multiple projects. However, the Secret Manager tool does not encrypt the stored secrets and should not be treated as a trusted store.
+
+For environments that support certificates and when developing a .NET Framework app, you can authenticate to Azure Key Vault with an X.509 certificate, whose private key would be managed by the OS. For more information on this approach, see [Authenticate with a Certificate instead of a Client Secret](https://docs.microsoft.com/azure/key-vault/key-vault-use-from-web-application#authenticate-with-a-certificate-instead-of-a-client-secret) and use the `AddAzureKeyVault()` overload that accepts an `X509Certificate2`.
```

Comment:
Is this better?

When developing a .NET Framework app in an environments that support certificates, you can authenticate to Azure Key Vault with an X.509 certificate. The X.509 certificates private key is managed by the OS. For more information ~~on this approach~~, see [Authenticate with a Certificate instead of a Client Secret](https://docs.microsoft.com/azure/key-vault/key-vault-use-from-web-application#authenticate-with-a-certificate-instead-of-a-client-secret). Use the `AddAzureKeyVault` overload that accepts an `X509Certificate2` with this approach.